internal/incident/incident.go

```diff
@@ -285,11 +276,18 @@ func (i *Incident) evaluateRules(eventID int64, causedBy types.Int) (types.Int,
 				Type:                      RuleMatched,
 				CausedByIncidentHistoryID: causedBy,
 			}
-			insertedID, err := i.AddRuleMatchedHistory(r, history)
+			err := i.AddRuleMatched(r)
+			if err != nil {
+				i.logger.Errorw("Failed to upsert incident rule", zap.String("rule", r.Name), zap.Error(err))
+
+				return types.Int{}, errors.New("failed to insert incident rule matched history")
+			}
+
+			insertedID, err := i.AddHistory(history, true)
```

Comment:
Also, as that was now moved out of the function, better move the initialization of the `HistoryRow` right above the call to `AddHistory()`.